indev/libinput.c

```diff
@@ -410,50 +550,79 @@ static void reset_scanned_devices(void) {
+      pointer_event = libinput_event_get_pointer_event(event);
+      break;
+    default:
+      return; /* We don't care about this events */
+  }
 
   /* We need to read unrotated display dimensions directly from the driver because libinput won't account
    * for any rotation inside of LVGL */
   lv_disp_drv_t *drv = lv_disp_get_default()->driver;
 
+  /* ignore more than 2 fingers as it will only confuse LVGL */
+  if (touch_event && (slot = libinput_event_touch_get_slot(touch_event)) > 1)
+    return;
+
+  evt = new_event(state);
+
   switch (type) {
     case LIBINPUT_EVENT_TOUCH_MOTION:
-    case LIBINPUT_EVENT_TOUCH_DOWN:
-      touch_event = libinput_event_get_touch_event(event);
-      lv_coord_t x_touch = libinput_event_touch_get_x_transformed(touch_event, drv->physical_hor_res > 0 ? drv->physical_hor_res : drv->hor_res) - drv->offset_x;
-      lv_coord_t y_touch = libinput_event_touch_get_y_transformed(touch_event, drv->physical_ver_res > 0 ? drv->physical_ver_res : drv->ver_res) - drv->offset_y;
-      if (x_touch < 0 || x_touch > drv->hor_res || y_touch < 0 || y_touch > drv->ver_res) {
+    case LIBINPUT_EVENT_TOUCH_DOWN: {
+      lv_coord_t x = libinput_event_touch_get_x_transformed(touch_event, drv->physical_hor_res > 0 ? drv->physical_hor_res : drv->hor_res) - drv->offset_x;
+      lv_coord_t y = libinput_event_touch_get_y_transformed(touch_event, drv->physical_ver_res > 0 ? drv->physical_ver_res : drv->ver_res) - drv->offset_y;
+      if (x < 0 || x > drv->hor_res || y < 0 || y > drv->ver_res) {
         break; /* ignore touches that are out of bounds */
       }
-      state->most_recent_touch_point.x = x_touch;
-      state->most_recent_touch_point.y = y_touch;
-      state->button = LV_INDEV_STATE_PR;
+      evt->point.x = x;
+      evt->point.y = y;
+      evt->pressed = LV_INDEV_STATE_PR;
+      evt->slot = slot;
+      state->slots[slot].point = evt->point;
+      state->slots[slot].pressed = evt->pressed;
       break;
+    }
     case LIBINPUT_EVENT_TOUCH_UP:
-      state->button = LV_INDEV_STATE_REL;
+      evt->pressed = LV_INDEV_STATE_REL;
+      state->slots[slot].pressed = evt->pressed;
+      evt->slot = slot;
       break;
     case LIBINPUT_EVENT_POINTER_MOTION:
-      pointer_event = libinput_event_get_pointer_event(event);
-      state->most_recent_touch_point.x += libinput_event_pointer_get_dx(pointer_event);
-      state->most_recent_touch_point.y += libinput_event_pointer_get_dy(pointer_event);
-      state->most_recent_touch_point.x = LV_CLAMP(0, state->most_recent_touch_point.x, drv->hor_res - 1);
-      state->most_recent_touch_point.y = LV_CLAMP(0, state->most_recent_touch_point.y, drv->ver_res - 1);
+      evt->point.x += libinput_event_pointer_get_dx(pointer_event);
+      evt->point.y += libinput_event_pointer_get_dy(pointer_event);
+      evt->point.x = LV_CLAMP(0, evt->point.x, drv->hor_res - 1);
+      evt->point.y = LV_CLAMP(0, evt->point.y, drv->ver_res - 1);
+      evt->is_relative = true;
```

Comment:
One more issue that I just found is that the code currently doesn't support dragging anymore. E.g. if you click and hold but even just slightly move the cursor, that triggers a release because we're not setting `evt->pressed` here.

Unfortunately, `libinput_event_pointer_get_button_state(..)` doesn't seem to work for motion events. So we cannot use it here.

We could maybe define two relativity flags, `is_relative_position` (which would mean add up `x` and `y` to the latest state) and `is_relative_button` (which would mean inherit the `pressed` value from the latest state). `LIBINPUT_EVENT_POINTER_MOTION` events would then require both `is_relative_position` and `is_relative_button` while  `LIBINPUT_EVENT_POINTER_BUTTON` events would only require `is_relative_position`.

Does that make sense?

Reply:
I haven't tested it but I just added a `pointer_button_down` property to `libinput_drv_state_t`, had that attribute be updated on `LIBINPUT_EVENT_POINTER_BUTTON` and set `evt->pressed` based on that. I also made `LIBINPUT_EVENT_POINTER_BUTTON` fallthrough to `LIBINPUT_EVENT_POINTER_MOTION` so that `evt->point` is set.

I think that should work?

Reply:
Nice! Yes, that seems to work. This now actually makes me think we could do the same for the delta movements. Instead of doing the weird `is_relative` thing when consuming events, just store last position we've seen in `state` and add the delta from libinput to it.

Reply:
Have made it so.